File: frontend/src/lib/parser/fuzzyMatch.ts
/**
 * Calculate Levenshtein distance between two strings
 */
export function levenshteinDistance(a: string, b: string): number {
  const matrix: number[][] = [];

  for (let i = 0; i <= b.length; i++) {
    matrix[i] = [i];
  }

  for (let j = 0; j <= a.length; j++) {
    matrix[0][j] = j;
  }

  for (let i = 1; i <= b.length; i++) {
    for (let j = 1; j <= a.length; j++) {
      if (b.charAt(i - 1) === a.charAt(j - 1)) {
        matrix[i][j] = matrix[i - 1][j - 1];
      } else {
        matrix[i][j] = Math.min(
          matrix[i - 1][j - 1] + 1, // substitution
          matrix[i][j - 1] + 1,     // insertion
          matrix[i - 1][j] + 1      // deletion
        );
      }
    }
  }

  return matrix[b.length][a.length];
}

/**
 * Check if two strings are fuzzy matches (Levenshtein distance <= maxDistance)
 */
export function isFuzzyMatch(input: string, target: string, maxDistance = 3): boolean {
  const inputLower = input.toLowerCase();
  const targetLower = target.toLowerCase();

  if (inputLower === targetLower) return true;

  return levenshteinDistance(inputLower, targetLower) <= maxDistance;
}

/**
 * Find best fuzzy match from a list of options
 */
export function findBestMatch(
  input: string,
  options: string[],
  maxDistance = 3
): { match: string; distance: number } | null {
  const inputLower = input.toLowerCase();
  let bestMatch: string | null = null;
  let bestDistance = Infinity;

  for (const option of options) {
    const distance = levenshteinDistance(inputLower, option.toLowerCase());
    if (distance < bestDistance && distance <= maxDistance) {
      bestMatch = option;
      bestDistance = distance;
    }
  }

  return bestMatch ? { match: bestMatch, distance: bestDistance } : null;
}

File: backend/app/services/parser/fuzzy_match.py
"""
Fuzzy string matching using Levenshtein distance.
"""

from typing import List, Optional, Tuple


def levenshtein_distance(a: str, b: str) -> int:
    """Calculate Levenshtein distance between two strings."""
    if len(a) < len(b):
        return levenshtein_distance(b, a)

    if len(b) == 0:
        return len(a)

    previous_row = list(range(len(b) + 1))

    for i, c1 in enumerate(a):
        current_row = [i + 1]
        for j, c2 in enumerate(b):
            # j+1 instead of j since previous_row and current_row are one character longer
            insertions = previous_row[j + 1] + 1
            deletions = current_row[j] + 1
            substitutions = previous_row[j] + (c1 != c2)
            current_row.append(min(insertions, deletions, substitutions))
        previous_row = current_row

    return previous_row[-1]


def is_fuzzy_match(input_str: str, target: str, max_distance: int = 3) -> bool:
    """Check if two strings are fuzzy matches (Levenshtein distance <= max_distance)."""
    input_lower = input_str.lower()
    target_lower = target.lower()

    if input_lower == target_lower:
        return True

    return levenshtein_distance(input_lower, target_lower) <= max_distance


def find_best_match(
    input_str: str, options: List[str], max_distance: int = 3
) -> Optional[Tuple[str, int]]:
    """Find best fuzzy match from a list of options."""
    input_lower = input_str.lower()
    best_match: Optional[str] = None
    best_distance = float("inf")

    for option in options:
        distance = levenshtein_distance(input_lower, option.lower())
        if distance < best_distance and distance <= max_distance:
            best_match = option
            best_distance = distance

    if best_match is not None:
        return (best_match, int(best_distance))
    return None
